drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,455 @@
+		 * cause this loop to spin without yielding to lower priority
+		 * threads until data is received.
+		 */
+		k_sleep(10);
+
+		/* Poll the socket for incoming data. */
+		err = poll(poll_fds, nfds, K_FOREVER);
+		if (err < 0) {
+			LOG_ERR("Poll error: %d", err);
+			continue;
+		} else if (err == 0) {
+			LOG_DBG("Poll timed out.");
+			continue;
+		}
+
+		k_mutex_lock(&common_socket_mutex, K_FOREVER);
+
+		bytes_read = recv(common_socket_fd, buf,
+				  sizeof(buf), MSG_DONTWAIT);
+
+		k_mutex_unlock(&common_socket_mutex);
+
+		if (bytes_read <= 0) {
+			/* Current implementation of poll() will cause us to
+			 * get here many times before receiving data.
+			 */
+			continue;
+		}
+
+		if (is_cesq_notification(buf, bytes_read)) {
+			cesq_notify(buf, bytes_read);
+			continue;
+		}
+
+		/* If it's not a notification, find out which client has a
+		 * pending command, and dispatch the response to it.
+		 */
+		client_idx = client_cmd_pending_get_idx();
+		if (client_idx == INVALID_CLIENT_IDX) {
+			LOG_DBG("No client with pending command.");
+			continue;
+		}
+
+		client = clients[client_idx];
+
+		memcpy(client->response, buf, bytes_read);
+		client->response_len = bytes_read;
+
+		/* Dispatch handler through client's work. */
+		LOG_DBG("\t <- SUBMITTING WORK TO CLIENT %d", client->id);
+		k_work_submit(&client->work);
+
+		done = cmd_is_done(client->response, bytes_read);
+		if (done) {
+			k_poll_signal_raise(&cmd_done_signal, 1);
+			LOG_DBG("\t <- SIGNALED THAT COMMAND IS PROCESSED");
+		}
+	}
+}
+
+static void at_cmd_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	struct at_cmd_fifo_item *fifo_cmd_data = NULL;
+	struct at_cmd_msg *cmd = NULL;
+	struct at_cmd_client *client = NULL;
+	int bytes_sent;
+	int err;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	struct k_poll_event fifo_event[1] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+					 K_POLL_MODE_NOTIFY_ONLY,
+					 &at_cmd_fifo)
+	};
+
+	struct k_poll_event signal_event[1] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
+					 K_POLL_MODE_NOTIFY_ONLY,
+					 &cmd_done_signal)
+	};
+
+	for (;;) {
+		err = k_poll(fifo_event, 1, K_FOREVER);
+		if (err) {
+			LOG_ERR("k_poll returned error: %d", err);
+			continue;
+		}
+
+		if (fifo_event[0].state == K_POLL_STATE_FIFO_DATA_AVAILABLE) {
+			fifo_cmd_data = k_fifo_get(&at_cmd_fifo, K_FOREVER);
+		} else {
+			continue;
+		}
+
+		if (fifo_cmd_data == NULL) {
+			continue;
+		}
+
+		fifo_event[0].state = K_POLL_STATE_NOT_READY;
+
+		cmd = &fifo_cmd_data->cmd;
+		client = cmd->client;
+		client->cmd_pending = true;
+
+		k_mutex_lock(&common_socket_mutex, K_FOREVER);
+
+		bytes_sent = send(common_socket_fd, cmd->buf, cmd->len, 0);
+
+		k_mutex_unlock(&common_socket_mutex);
+
+		if (bytes_sent <= 0) {
+			LOG_ERR("Could not send AT command to modem: %d",
+								bytes_sent);
+			goto error;
+		}
+
+		if (is_cesq_subscribe(cmd->buf, cmd->len)) {
```

Comment:
To disable CESQ notifications you send ```AT%CESQ=0```.

Reply:
Yes, but it's not my question. Because if you issue an AT%CESQ=1 the client will be put into cesq_subscribers list, but there is no way to remove the client from that list if I don't want the CESQ notification. There may be other clients still wanting the CESQ notifications so just sending an AT%CESQ=0 will also not work.

Then you will need to remove the client that issued the CESQ command from that list, and then check if there are other clients that still subscribe, and if not issue the command to the modem.

Reply:
That actually got me to think about another problem, because how should state be negosiated between the clients? So AT%CESQ has only two states, but for instance AT+CFUN has 4 or AT+CEREG has 6 (at least that is what the nrf91 AT Commands document states). How to decide what state to select if two clients disagree?

Reply:
What you describe @evenl was my initial idea for a naïve implementation. The best alternative that I see is using the event manager module, but zero experience with that made me just implement this first for testing.

Reply:
> That actually got me to think of another problem, because how should state be negotiated between the clients? So AT%CESQ has only two states (but still applicable), but for instance AT+CFUN has 4 or AT+CEREG has 6 (at least that is what the nrf91 AT Commands document states). How to decide what state to select if two clients disagree?

Good point. The easy way is to leave handling of the link itself to the LTE link controller, and not allow certain commands here.

Reply:
Ok, so that is option one, but that is the same as saying that we don't take care of it and trust that the different clients behaves, which that is a valid option. But I think I would like to either get some feedback or have some way to check if my wanted state don't conflict with any other clients requered state.

Reply:
We could return a descriptive error code if a client attempts to use commands on a "blacklist" or something. Haven't really thought about this before, it's an issue with the entire system as it is now. Any module can put the link/modem in a state that other parts of the system don't expect or can't function properly under. Having all parts use the same interface would at least make it clearer when such conflicting state requests occur.